common/quotas/clocked_rate_limiter.go

```diff
@@ -0,0 +1,154 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package quotas
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+	"golang.org/x/time/rate"
+)
+
+// ClockedRateLimiter wraps a rate.Limiter with a clockwork.Clock. It is used to ensure that the rate limiter respects
+// the time determined by the clock.
+type ClockedRateLimiter struct {
+	rateLimiter *rate.Limiter
+	clock       clockwork.Clock
+}
+
+var (
+	ErrRateLimiterWaitInterrupted                       = errors.New("rate limiter wait interrupted")
+	ErrRateLimiterReservationCannotBeMade               = errors.New("rate limiter reservation cannot be made due to insufficient quota")
+	ErrRateLimiterReservationWouldExceedContextDeadline = errors.New("rate limiter reservation would exceed context deadline")
+)
+
+func NewClockedRateLimiter(rateLimiter *rate.Limiter, clock clockwork.Clock) ClockedRateLimiter {
+	return ClockedRateLimiter{
+		rateLimiter: rateLimiter,
+		clock:       clock,
+	}
+}
+
+func (l ClockedRateLimiter) Allow() bool {
+	return l.AllowN(l.clock.Now(), 1)
+}
+
+func (l ClockedRateLimiter) AllowN(now time.Time, token int) bool {
+	return l.rateLimiter.AllowN(now, token)
+}
+
+// ClockedReservation wraps a rate.Reservation with a clockwork.Clock. It is used to ensure that the reservation
+// respects the time determined by the clock.
+type ClockedReservation struct {
```

Comment:
nit: I would create a separate file for clockedReservation impl, but no strong opinion.

Reply:
I normally would also, but I'm keeping it here because it's only useful in the context of the rate limiter, and it would be weird for it to have its own test file, too.